python/ray/tune/utils/placement_groups.py

```diff
@@ -0,0 +1,227 @@
+    so trials that should be run before other trials whose placement
+    groups are not ready, yet, can cut the queue.
+    """
+
+    def __init__(self):
+        self._staging: Dict[Trial, PlacementGroup] = {}
+        self._ready: Dict[Trial, PlacementGroup] = {}
+        self._staging_futures: Dict[ObjectRef, Trial] = {}
+        self._in_use: Set[PlacementGroup] = set()
+
+        self._latest_staging_start_time = time.time()
+
+    @property
+    def available_trials(self):
+        """Get trials with a ready placement group that have not been started.
+
+        Yields: Trial objects.
+
+        """
+        for ready, pg in self._ready.items():
+            if pg in self._in_use:
+                continue
+            yield ready
+
+    def stage_trial(self, trial: Trial):
+        """Stage a trial placement group.
+
+        Create the trial placement group if maximum number of pending
+        placement groups is not exhausted.
+
+        Args:
+            trial (Trial): Trial to stage.
+
+        Returns: False if trial has not been staged, True otherwise.
+
+        Creates its placement group and moves it to `self._staging`.
+        """
+        if trial in self._staging:
+            raise RuntimeError(f"Trial {trial} already staged.")
+
+        if not self.can_stage():
+            return False
+
+        pg = trial.placement_group_factory()
+
+        self._staging[trial] = pg
+        self._staging_futures[pg.ready()] = trial
+
+        self._latest_staging_start_time = time.time()
+
+        return True
+
+    def can_stage(self):
+        """Return True if we can stage another placement group."""
+        return len(self._staging) < TUNE_MAX_PENDING_TRIALS_PG
+
+    def update_status(self):
+        """Update placement group status.
+
+        Moves ready placement groups from `self._staging` to
+        `self._ready`.
+        """
+        ready = True
+        while ready:
+            # Use a loop as `ready` might return futures one by one
+            ready, _ = ray.wait(list(self._staging_futures.keys()), timeout=0)
+
+            for ready_fut in ready:
+                ready_trial = self._staging_futures.pop(ready_fut)
+                ready_pg = self._staging.pop(ready_trial)
+                self._ready[ready_trial] = ready_pg
+
+    def get_full_actor_cls(self, trial: Trial,
+                           actor_cls: ActorClass) -> Optional[ActorClass]:
+        """Get a fully configured actor class.
+
+        Returns the actor handle if the placement group is ready.
+
+        Args:
+            trial (Trial): trial object to start
+            actor_cls: Ray actor class.
+
+        Returns: Configured ActorClass or None
+
+        """
+        if trial not in self._ready:
+            return None
+
+        pg = self._ready[trial]
+        self._in_use.add(pg)
+
+        return actor_cls.options(
+            placement_group=pg, placement_group_bundle_index=0)
```

Comment:
wait @rkooo567 so if my actor_cls needs a GPU, then I must provide the GPU resource here too right?

Reply:
I'm happy btw to also pass resources, but we'll need to think about the API for this (as we're currently using `resources_per_trial` for the placement group factory)

Reply:
Maybe as a starter, we can just have the head actor take all the resources of bundle 0?